File: clients/drive9-js/src/patch.ts
import { Client } from "./client.js";
import { bodyInit, bufferSource } from "./compat.js";
import { checkError, Drive9Error } from "./error.js";
import type { PatchPartURL, PatchPlan } from "./models.js";

const DEFAULT_PART_SIZE = 8 * 1024 * 1024;

export type ReadPartFn = (partNumber: number, partSize: number, origData?: Uint8Array) => Uint8Array;
export type ProgressFn = (uploaded: number, total: number) => void;

export async function patchFileImpl(
  client: Client,
  path: string,
  newSize: number,
  dirtyParts: number[],
  readPart: ReadPartFn,
  progress?: ProgressFn,
  partSize = DEFAULT_PART_SIZE
): Promise<void> {
  const resp = await fetch(client.fsUrl(path), {
    method: "PATCH",
    headers: client["authHeaders"]({ "Content-Type": "application/json" }),
    body: JSON.stringify({ new_size: newSize, dirty_parts: dirtyParts, part_size: partSize }),
  });
  await checkError(resp);
  const plan = (await resp.json()) as PatchPlan;
  const total = plan.upload_parts.reduce((sum, p) => sum + p.size, 0);
  let uploaded = 0;

  const parallelism = 16;
  const semaphore = new Semaphore(parallelism);
  const tasks: Promise<void>[] = [];

  for (const part of plan.upload_parts) {
    tasks.push(
      (async () => {
        await semaphore.acquire();
        try {
          await uploadPatchPart(client, part, readPart);
          uploaded += part.size;
          if (progress) progress(uploaded, total);
        } finally {
          semaphore.release();
        }
      })()
    );
  }
  await Promise.all(tasks);
}

async function uploadPatchPart(client: Client, part: PatchPartURL, readPart: ReadPartFn): Promise<void> {
  let origData: Uint8Array | undefined;
  if (part.read_url) {
    const headers: Record<string, string> = {};
    if (part.read_headers) {
      for (const [k, v] of Object.entries(part.read_headers)) {
        if (typeof v === "string") headers[k] = v;
      }
    }
    const resp = await fetch(part.read_url, { headers });
    await checkError(resp);
    origData = new Uint8Array(await resp.arrayBuffer());
  }
  const data = readPart(part.number, part.size, origData);
  const checksum = await sha256Base64(data);
  const headers: Record<string, string> = { "x-amz-checksum-sha256": checksum };
  if (part.headers) {
    for (const [k, v] of Object.entries(part.headers)) {
      if (typeof v === "string" && k.toLowerCase() !== "host") {
        headers[k] = v;
      }
    }
  }
  const resp = await fetch(part.url, { method: "PUT", headers, body: bodyInit(data) });
  await checkError(resp);
}

async function sha256Base64(data: Uint8Array): Promise<string> {
  const hash = await crypto.subtle.digest("SHA-256", bufferSource(data));
  return Buffer.from(hash).toString("base64");
}

class Semaphore {
  private permits: number;
  private waiters: (() => void)[] = [];
  constructor(n: number) {
    this.permits = n;
  }
  acquire(): Promise<void> {
    if (this.permits > 0) {
      this.permits--;
      return Promise.resolve();
    }
    return new Promise((resolve) => this.waiters.push(resolve));
  }
  release(): void {
    const next = this.waiters.shift();
    if (next) {
      next();
    } else {
      this.permits++;
    }
  }
}

File: clients/drive9-rs/src/patch.rs
use crate::client::Client;
use crate::error::{check_error, Drive9Error};
use crate::models::PatchPartURL;
use reqwest::header::{HeaderMap, HeaderName, CONTENT_TYPE};
use serde_json::json;
use sha2::{Digest, Sha256};
use std::sync::Arc;
use tokio::sync::Semaphore;

impl Client {
    pub async fn patch_file<F>(
        &self,
        path: &str,
        new_size: i64,
        dirty_parts: &[i32],
        read_part: F,
    ) -> Result<(), Drive9Error>
    where
        F: Fn(i32, i64, Option<&[u8]>) -> Result<Vec<u8>, Drive9Error> + Send + Sync + 'static,
    {
        let body = json!({
            "new_size": new_size,
            "dirty_parts": dirty_parts,
        });
        let resp = self
            .http
            .patch(self.fs_url(path))
            .headers(self.auth_headers())
            .header(CONTENT_TYPE, "application/json")
            .body(body.to_string())
            .send()
            .await?;
        let resp = check_error(resp).await?;
        let plan: crate::models::PatchPlan = resp.json().await?;

        let _total_parts = plan.upload_parts.len() + plan.copied_parts.len();
        let sem = Arc::new(Semaphore::new(4));
        let mut tasks = vec![];
        let read_part = Arc::new(read_part);

        for part in plan.upload_parts {
            let permit = Arc::clone(&sem).acquire_owned().await.unwrap();
            let client = self.clone();
            let rf = Arc::clone(&read_part);
            let task = tokio::spawn(async move {
                let _permit = permit;
                client.upload_patch_part(&part, rf.as_ref()).await
            });
            tasks.push(task);
        }

        let mut first_err = None;
        for t in tasks {
            match t.await {
                Ok(Err(e)) if first_err.is_none() => first_err = Some(e),
                Err(e) if first_err.is_none() => {
                    first_err = Some(Drive9Error::Other(format!("task panicked: {}", e)))
                }
                _ => {}
            }
        }
        if let Some(e) = first_err {
            return Err(e);
        }

        self.complete_upload(&plan.upload_id).await
    }

    async fn upload_patch_part<F>(
        &self,
        part: &PatchPartURL,
        read_part: &F,
    ) -> Result<(), Drive9Error>
    where
        F: Fn(i32, i64, Option<&[u8]>) -> Result<Vec<u8>, Drive9Error> + Send + Sync + 'static,
    {
        let orig_data = if let Some(ref read_url) = part.read_url {
            let mut headers = HeaderMap::new();
            if let Some(ref rh) = part.read_headers {
                for (k, v) in rh {
                    if let Ok(hv) = reqwest::header::HeaderValue::from_str(v.as_str().unwrap_or(""))
                    {
                        headers.insert(
                            HeaderName::from_bytes(k.as_bytes()).unwrap_or(CONTENT_TYPE),
                            hv,
                        );
                    }
                }
            }
            let resp = self.http.get(read_url).headers(headers).send().await?;
            let resp = check_error(resp).await?;
            Some(resp.bytes().await?.to_vec())
        } else {
            None
        };

        let data = read_part(part.number, part.size, orig_data.as_deref())?;
        let checksum = base64::Engine::encode(
            &base64::engine::general_purpose::STANDARD,
            Sha256::digest(&data),
        );

        let mut headers = HeaderMap::new();
        headers.insert(
            "x-amz-checksum-sha256",
            reqwest::header::HeaderValue::from_str(&checksum).unwrap(),
        );
        if let Some(ref ph) = part.headers {
            for (k, v) in ph {
                if let Ok(hv) = reqwest::header::HeaderValue::from_str(v.as_str().unwrap_or("")) {
                    if k.eq_ignore_ascii_case("host") {
                        continue;
                    }
                    headers.insert(
                        HeaderName::from_bytes(k.as_bytes()).unwrap_or(CONTENT_TYPE),
                        hv,
                    );
                }
            }
        }
        let resp = self
            .http
            .put(&part.url)
            .headers(headers)
            .body(data)
            .send()
            .await?;
        check_error(resp).await?;
        Ok(())
    }
}
